rust/src/core.rs

```diff
@@ -13,27 +14,27 @@ pub struct TempEtcGuard {
     renamed_etc: bool,
 }
 
-impl TempEtcGuard {
-    /// Create a context with a temporary /etc, and return a guard to it.
-    pub fn undo_usretc(rootfs: openat::Dir) -> anyhow::Result<Self> {
-        let has_usretc = rootfs.exists("usr/etc")?;
-        if has_usretc {
-            // In general now, we place contents in /etc when running scripts
-            rootfs.local_rename("usr/etc", "etc")?;
-            // But leave a compat symlink, as we used to bind mount, so scripts
-            // could still use that too.
-            rootfs.symlink("usr/etc", "../etc")?;
-        }
-
-        let guard = Self {
-            rootfs,
-            renamed_etc: has_usretc,
-        };
-        Ok(guard)
+pub fn prepare_tempetc_guard(rootfs: i32) -> Result<Box<TempEtcGuard>> {
+    let rootfs = ffi_view_openat_dir(rootfs);
+    let has_usretc = rootfs.exists("usr/etc")?;
+    let mut renamed_etc = false;
+    if has_usretc {
+        // In general now, we place contents in /etc when running scripts
+        rootfs.local_rename("usr/etc", "etc")?;
+        // But leave a compat symlink, as we used to bind mount, so scripts
+        // could still use that too.
+        rootfs.symlink("usr/etc", "../etc")?;
+        renamed_etc = true;
     }
+    Ok(Box::new(TempEtcGuard {
+        rootfs,
+        renamed_etc,
+    }))
+}
 
+impl TempEtcGuard {
     /// Remove the temporary /etc, and destroy the guard.
-    pub fn redo_usretc(self) -> anyhow::Result<()> {
+    pub fn undo(&self) -> anyhow::Result<()> {
```

Comment:
Would be cool if we could we make the `undo` happen in `Drop::drop` so we don't have to explicitly call it. This matches some other guards we have in C which used autocleanups to "unguard".

Reply:
We could do that actually!  But I explicitly chose not to because there's no way to check for errors in `drop()`.  (This gets into the whole rationale behind things like https://docs.rs/drop_bomb/0.1.5/drop_bomb/ that we use in https://docs.rs/openat-ext/0.1.9/openat_ext/struct.FileWriter.html )